File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.ans = 0
    def dia(self, root: Optional[TreeNode]):
        if root == None:
            return 0
        left = self.dia(root.left)
        right = self.dia(root.right)
        self.ans = max(self.ans, left+right)
        return 1 + max(left,right)
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.dia(root)
        return self.ans
        

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp

class Solution {
   int dia(TreeNode* root, int& ans)
    {
        if(root==NULL) return 0;

        int left = dia(root->left,ans);
        int right = dia(root->right,ans);
        ans = max(ans,left+right);
        return 1 + max(left,right);
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int ans = 0;
        int hei = dia(root,ans);
        return ans;
    }
};